prov/efa/test/efa_unit_test_cq.c

```diff
@@ -0,0 +1,249 @@
+
+	memcpy(raw_addr.raw, g_ibv_gid_mock, EFA_GID_LEN);
+	raw_addr.qpn = 0;
+	raw_addr.qkey = 1234;
+	ret = fi_av_insert(resource.av, &raw_addr, 1, &addr, 0 /* flags */, NULL /* context */);
+	assert_int_equal(ret, 1);
+
+	err = fi_send(resource.ep, send_buff.buff, send_buff.size, fi_mr_desc(send_buff.mr), addr, NULL /* context */);
+	assert_int_equal(err, 0);
+
+	ret = fi_cq_read(resource.cq, &cq_entry, 1);
+	assert_int_equal(ret, -FI_EAVAIL);
+
+	ret = fi_cq_readerr(resource.cq, &cq_err_entry, 0);
+	assert_int_equal(ret, 1);
+
+	if (ep_type == FI_EP_DGRAM) {
+		assert_int_equal(cq_err_entry.err, FI_EIO);
+	} else {
+		assert_int_equal(cq_err_entry.err, IBV_WC_GENERAL_ERR);
+	}
+	assert_int_equal(cq_err_entry.prov_errno, IBV_WC_GENERAL_ERR);
+
+	efa_unit_test_buff_destruct(&send_buff);
+	efa_unit_test_resource_destruct(&resource);
+}
+
+/**
+ * @brief test that RDM CQ's fi_cq_read()/fi_cq_readerr() works properly when rdma-core return bad status for send.
+ *
+ * When ibv_post_send() operation failed, fi_cq_read() should return -FI_EAVAIL, which means error available.
+ * then user should call fi_cq_readerr() to get an error CQ entry that contain error code.
+ */
+void test_rdm_cq_read_bad_send_status()
+{
+	test_cq_read_bad_send_status(FI_EP_RDM);
+}
+
+/**
+ * @brief test that DGRAM CQ's fi_cq_read()/fi_cq_readerr() works properly when rdma-core return bad status for send.
+ *
+ * When ibv_post_send() operation failed, fi_cq_read() should return -FI_EAVAIL, which means error available.
+ * then user should call fi_cq_readerr() to get an error CQ entry that contain error code.
+ */
+void test_dgram_cq_read_bad_wc_status()
+{
+	test_cq_read_bad_send_status(FI_EP_DGRAM);
+}
+
+/**
+ * @brief verify that fi_cq_read/fi_cq_readerr works properly when rdma-core return bad status for recv.
+ *
+ * When an ibv_post_recv() operation failed, no data was received. Therefore libfabric cannot
+ * find the corresponding RX operation to write a CQ error. It will write an EQ error instead.
+ */
+void test_rdm_cq_read_bad_recv_status()
+{
+	struct efa_cq *efa_cq;
+	struct rxr_ep *rxr_ep;
+	struct efa_resource resource = {0};
+	struct rxr_pkt_entry *pkt_entry;
+	struct fi_cq_data_entry cq_entry;
+	struct fi_eq_err_entry eq_err_entry;
+	int ret;
+
+	efa_unit_test_resource_construct(&resource, FI_EP_RDM);
+	rxr_ep = container_of(resource.ep, struct rxr_ep, util_ep.ep_fid);
+
+	pkt_entry = rxr_pkt_entry_alloc(rxr_ep, rxr_ep->efa_rx_pkt_pool, RXR_PKT_FROM_EFA_RX_POOL);
+	assert_non_null(pkt_entry);
+
+	efa_cq = container_of(rxr_ep->rdm_cq, struct efa_cq, util_cq.cq_fid);
+	efa_cq->ibv_cq_ex->start_poll = &efa_mock_ibv_start_poll_return_mock;
+	efa_cq->ibv_cq_ex->end_poll = &efa_mock_ibv_end_poll_check_mock;
+	efa_cq->ibv_cq_ex->read_opcode = &efa_mock_ibv_read_opcode_return_mock;
+
+	will_return(efa_mock_ibv_start_poll_return_mock, 0);
+	will_return(efa_mock_ibv_end_poll_check_mock, NULL);
+	will_return(efa_mock_ibv_read_opcode_return_mock, IBV_WC_RECV);
```

Comment:
This mock is unnecessary.

Reply:
Why?

Reply:
Because right after this there is `will_return_maybe(efa_mock_ibv_read_opcode_return_mock, IBV_WC_RECV);`

Reply:
I think I need to use `will_return_always` because I want to make sure it is called at least once.